[Core] ProgressBar Handlers

jsuarezruiz opened the pull request:
### Description of Change ###

ProgressBar Handlers.

| API | Android | iOS | 
| ----|:--------:|:---:|
| Progress | ✅  | ✅  |
| ProgressColor | ✅  | ✅  | 
| ProgressTo | ✅  | ✅  | 

### Platforms Affected ### 

- Core
- iOS
- Android

### PR Checklist ###

- [x] Targets the correct branch
- [x]  Include Handler tests
- [ ] Tests are passing (or failures are unrelated)

hartez commented on src/Platform.Handlers/src/Xamarin.Platform.Handlers/Core/IProgress.cs:
```diff
@@ -0,0 +1,13 @@
+		double Progress { get; }
+		Color ProgressColor { get; }
+
+		Task<bool> ProgressTo(double value, uint length, Easing easing);
```
"Easing" seems like an animation implementation detail; should that parameter be at the interface level?

jsuarezruiz replied:
Right, removed.

hartez commented on src/Platform.Handlers/src/Xamarin.Platform.Handlers/Core/IProgress.cs:
```diff
@@ -0,0 +1,13 @@
+
+namespace Xamarin.Platform
+{
+	public interface IProgress : IView
```
Got some spaces/tabs issues here.

jsuarezruiz commented:
I have rebased and moved the macOS implementation to: https://github.com/xamarin/Xamarin.Forms/tree/progressbarhandler-mac in order to facilitate the revision and merge of this PR.

jsuarezruiz commented:
Added _DO-NOT-MERGE!!!_ tag until adding handler tests.

mattleibow commented on src/Platform.Handlers/src/Xamarin.Platform.Handlers/Core/IProgress.cs:
```diff
@@ -0,0 +1,10 @@
+	public interface IProgress : IView
+	{
+		double Progress { get; }
+		Color ProgressColor { get; }
```
ProgressColor or maybe something to do with foreground?

jsuarezruiz replied:
Do you mean use a Brush?. Have sense for me but probably not in this PR, after include Brushes support (P2-P3?).

mattleibow commented on src/Platform.Handlers/src/Xamarin.Platform.Handlers/Handlers/ProgressBar/ProgressBarHandler.cs:
```diff
@@ -0,0 +1,42 @@
+		{
+			[nameof(IProgress.Progress)] = MapProgress,
+			[nameof(IProgress.ProgressColor)] = MapProgressColor,
+#if __IOS__
+			[nameof(IView.BackgroundColor)] = MapBackgroundColor
+#endif
```
Why is this iOS only?

mattleibow replied:
Even if this is just for iOS, how will this translate to me when I am customizing the handler?

mattleibow replied:
I see Android has a progressBackgroundTint which may be this, but for android? We may want to override this.

jsuarezruiz replied:
Because in the ProgressBar Renderer we set the `TrackTintColor` property in iOS. BUT...this a good point. Now, in Xamarin.Forms the BackgroundColor property behavior is different between iOS and Android. Must we match the behavior (use `progressBackgroundTint`)?

jsuarezruiz replied:
Now, Android and iOS have the same behavior.

mattleibow commented on src/Platform.Handlers/src/Xamarin.Platform.Handlers/Platform/Android/ProgressBarExtensions.cs:
```diff
@@ -0,0 +1,42 @@
+	{
+		public static void UpdateProgress(this AProgressBar nativeProgressBar, IProgress progress)
+		{
+			nativeProgressBar.Progress = (int)(progress.Progress * 10000);
```
This number is big and scary. Needs to be a constant.

Also, this is 10K and the slider was 1K. Should this be the same, somehow?

jsuarezruiz replied:
Created a constant.

mattleibow commented on src/Platform.Handlers/src/Xamarin.Platform.Handlers/Platform/iOS/ProgressBarExtensions.cs:
```diff
@@ -0,0 +1,23 @@
+
+		public static void UpdateProgressColor(this UIProgressView nativeProgressBar, IProgress progress)
+		{
+			nativeProgressBar.ProgressTintColor = progress.ProgressColor == Color.Default ? null : progress.ProgressColor.ToNative();
+		}
+
+		public static void UpdateBackgroundColor(this UIProgressView nativeProgressBar, IProgress progress)
+		{
+			nativeProgressBar.TrackTintColor = progress.BackgroundColor != Color.Default ? progress.BackgroundColor.ToNative() : null;
```
NIT: Why is one == and the other !=?

mattleibow replied:
Maybe we can get a new extension method for ToNative that takes a param to check? `ToNative(Color? valueIfNull)` so this code can have reduced chance of someone mixing the order of results.

Redth commented on src/Platform.Handlers/src/Xamarin.Platform.Handlers/Handlers/ProgressBar/ProgressBarHandler.Android.cs:
```diff
@@ -0,0 +1,13 @@
+﻿using static Android.Resource;
+using AProgressBar = Android.Widget.ProgressBar;
```
A bit nitpicky but I think we should try and use more detailed aliases here as it took me a minute to figure out what this was.  Perhaps we can think of a convention around this `PlatformX`, or just even use the platform name `AndroidProgressBar`.

jsuarezruiz replied:
Renamed to `AndroidProgressBar`

mattleibow commented on src/Platform.Handlers/tests/Xamarin.Platform.Handlers.DeviceTests/Handlers/ProgressBar/ProgressBarTests.cs:
```diff
@@ -0,0 +1,32 @@
+		{
+			var progressBar = new ProgressBarStub()
+			{
+				Progress = 0.5
```
Can we make this a theory with a few values? Like 0, 0.3, 0.5, 0.75 and 1? Maybe even add things out of the range, like -1, 32 and -0.45. This way we can test more than just what we know works.

jsuarezruiz replied:
Changed the test to be a Theory with different values.

mattleibow commented on src/Platform.Handlers/tests/Xamarin.Platform.Handlers.DeviceTests/Handlers/ProgressBar/ProgressBarTests.cs:
```diff
@@ -0,0 +1,32 @@
+		{
+			var progressBar = new ProgressBarStub()
+			{
+				ProgressColor = Color.Red
```
Can this also be a theory to provide a few colors. At least the Red, Green, Blue and maybe the others like Yellow and Black and White. Single color tests I found might pass because of platform defaults. I know red is not common, but who knows what the future brings and iOS goes all red to celebrate Christmas Santas.

jsuarezruiz replied:
Good point!. Changed the test to be a Theory with different values.

jsuarezruiz commented:
I close this PR, it has been moved to dotnet/maui#418
